Wrote flutter_todos tutorial (for v8) (#3047)
felangel/bloc: docs/fluttertodostutorial.md

> 

Reviewer: ```suggestion
```
nit but can we remove the extra newline?

# Flutter Todos Tutorial

![advanced](https://img.shields.io/badge/level-advanced-red.svg)

> In the following tutorial, we're going to build a Todos App in Flutter using the Bloc library.

![demo](./assets/gifs/flutter_todos.gif)
> ZZTODO: Update gif with the movie in the issue tracker.

## Key Topics

- Observe state changes with [BlocObserver](/coreconcepts?id=blocobserver).
- [BlocProvider](/flutterbloccoreconcepts?id=blocprovider), Flutter widget which provides a bloc to its children.
- [BlocBuilder](/flutterbloccoreconcepts?id=blocbuilder), Flutter widget that handles building the widget in response to new states.
- [BlocListener](/flutterbloccoreconcepts?id=bloclistener), a Flutter widget which invokes the listener code in response to state changes in the bloc.

> - Using 3 Blocs and  1 Cubit. [What's the difference?](/coreconcepts?id=cubit-vs-bloc)

Reviewer: ```suggestion
- Using Blocs and Cubit. [What's the difference?](/coreconcepts?id=cubit-vs-bloc)
```

I feel the amount isn't that important and it could get outdated pretty quickly. wdyt?

Author: I could go either way, but I like keeping the numbers. Otherwise, the stuff at the top is easy to skip over.

> - Uses a [Repository and API layer](architecture/repository) separately for the data.

Reviewer 2: ```suggestion
- Uses a [Repository and API layer](architecture/repository) separately for the domain and data logic, respectively.
```

> - [RespositoryProvider](/flutterbloccoreconcepts?id=respositoryprovider), a flutter widget to provide a repository to its children.

Reviewer 2: ```suggestion
- [RepositoryProvider](/flutterbloccoreconcepts?id=respositoryprovider), a Flutter widget which provides a repository to its children.
```

- Prevent unnecessary rebuilds with [Equatable](/faqs?id=when-to-use-equatable).

> - [MultiBlocListener](/flutterbloccoreconcepts?id=multibloclistener), a Flutter widget that merges multiple BlocListener widgets into one.

Reviewer 2: ```suggestion
- [MultiBlocListener](/flutterbloccoreconcepts?id=multibloclistener), a Flutter widget that sets up multiple BlocListener widgets that allow us to react to different states.
```

Reviewer 2: Not sure about this one, it can be reformed better

Author: Let's change to:
... sets up ...... at once. OR
... sets up ...... at the same point in the widget tree.

The "allows us to react to different states" isn't unique to MultiBlocListener. It's just a convenience /shortcut.

BTW: can you @marcossevilla add me as a collaborator on this issue or ask felix to do so? Otherwise, I can't suggest the change directly, I think, in github. 

Add suggestion to batch is greyed out.

Reviewer: @chonghorizons you should be able to edit because it's your branch.

Author: @felangel Unfortunately, "Add suggestion to branch" is still greyed out:
![image](https://user-images.githubusercontent.com/36284839/146476453-34a2bc3f-400d-44d3-9157-20ae87202049.png)

I can edit the forked branch, but I think the issue is that I don't have write permission on this branch.

> 

Reviewer: ```suggestion
```

Author: Good catch.

## Setup

We'll start off by creating a brand new Flutter project

> ```shell

Reviewer: ```suggestion
```sh
```

minor but for consistency can we just `sh`?

> flutter create flutter_login

Reviewer: Can we revert this change? The previous script should still be correct and it should be `flutter create flutter_todos` not `flutter_login`.

Reviewer 2: I think we should mention that the project was built using `very_good_cli` to redirect them to that package documentation in case they have questions about specific features that the CLI sets for the project, wdyt?

```

We can then replace the contents of `pubspec.yaml` with

[pubspec.yaml](_snippets/flutter_todos_tutorial/pubspec.yaml.md ':include')

and then install all of the dependencies

> ```shell

Reviewer: Can we revert this change as well?

Author: Yes. This might be faster. I recommend making a list of final polish/styleguide steps here:

https://github.com/felangel/bloc/pull/3047#issuecomment-989334851

If you focus on formatting you'll never get through the tutorial.

It's faster to just write a list of style-guide items to implement (like a human-implemented css file) than it is to make the individual changes.

Take a look via docsify. There are notes to ignore in the raw markdown.

flutter pacakages get
```

> [comment]: <> (> ZTODO: Is below still true? I think so, but I don't know the history of this.)

Reviewer: No this shouldn't be the case anymore with the rewritten version.

[comment]: <> (> **Note:** We're overriding some dependencies because we're going to be reusing them from [Brian Egan's Flutter Architecture Samples]&#40;https://github.com/brianegan/flutter_architecture_samples&#41;.)

## Directory Structure

Because this is a long tutorial, presenting the directory structure is helpful.

> Alternatively, explore the files on github or clone them locally and open in your IDE. The full source for this example can be found [here](https://github.com/felangel/Bloc/tree/master/examples/flutter_todos).

> We will be using 3 blocs, 1 cubit, and 3 packages.

Reviewer 2: I think we should explain this in sections, @chonghorizons I can share with you the outline I had for this tutorial so we can just explain the components future-proofing the quantities we’re using right now in case we make the tutorial more complex, we just need to add a new section and not changing these numbers, wdyt?

Author: That would be okay, but I think the numbers are helpful for newcomers reading the docs. It's also pointing out to look at the directory structure to see this.

It could be more clear by breaking it out as a

> TIP: Look at the directory structure. By using a good naming convention and directory structure, we can easily see: We will be using 3 blocs, 1 cubit, and 3 packages.

Sorry, I can't edit directly since I'm not a collaborator.
